### ISSUE TITLE
About the instant API

### ISSUE THREAD

**Issue by GoogleCodeExporter** [javilechero/android-bluez-ime#76]
```
I am trying to integrate bluez with my app. I have to use the instant API for 
receiving the analog stick events and have figured out that all your APIs are 
based on Intents and event broadcasts. I don't think this is a very efficient 
approach. I wonder why you did not choose to use Service to provide APIs 
through RPC? After all, that is the official way on Android for an app to 
provide APIs to other apps. The author of the Wiimote Controller app is using 
this approach.

Thanks for listening.
```

Original issue reported on code.google.com by `freeman....@gmail.com` on 6 Sep 2011 at 10:13

**Comment by GoogleCodeExporter** [comment]
```
Sorry, I did not mean 'Defect', and now don't know how to change its type. 
Please if you can, change to Enhancement.
```

Original comment by `freeman....@gmail.com` on 6 Sep 2011 at 10:16

**Comment by GoogleCodeExporter** [comment]
```
Can you point me to an example that uses the service API? From what I read, the 
broadcast messages are an efficient communication method.
```

Original comment by `kenneth@hexad.dk` on 6 Sep 2011 at 10:19
- Added labels: **Type-Enhancement**
- Removed labels: **Type-Defect**

**Comment by GoogleCodeExporter** [comment]
```
I don't know a good example. But probably you can take a look at how this is 
implemented in the Wiimote controller app:
http://android.ccpcreations.com/wiicontroller/wc-direct-interface

I haven't done any performance test yet. Just thought it could be better.
```

Original comment by `freeman....@gmail.com` on 6 Sep 2011 at 12:02

**Comment by GoogleCodeExporter** [comment]
```
I can see that it is doable that way. I did not consider it because I find 
broadcast events so easy to use, and they can be implemented without relying on 
.aidl or other static information, just strings. 

Since broadcasts are used extensively in Android I am fairly certain that they 
are effective. I cannot see any performance measurements that compare the two, 
nor can I find any official documentation explaining when to use what. But 
since the IPC method also sends messages as parcels (underneath), I assume they 
will do just about the same. 

The broadcasts have the advantage that the OS schedules delivery, which means 
that multiple receivers can process the messages at the same time, and the 
reader is not blocked on sending. Still, that is just guessing at actual 
performance, so if someone does a test, I will consider implementing an IPC. If 
you would prefer to use an IPC, you are most welcome to add it, just send me a 
private email with your google account name, and I will grant you commit access 
to the project.
```

Original comment by `kenneth@hexad.dk` on 6 Sep 2011 at 12:51

**Comment by GoogleCodeExporter** [comment]
```
First I'll try to integrate your current API, and then measure it. If I feel a 
noticeable delay in the game, I'll consider contributing my code.

BTW, very appreciated that you'd want to grant me commit access to your 
excellent project :-) 
```

Original comment by `freeman....@gmail.com` on 6 Sep 2011 at 1:21

**Comment by GoogleCodeExporter** [comment]
```
I am done with the integration. I haven't yet noticed any delay on key 
dispatching :-)

My question now is: will your APIs be backward compatible? Do you recommend 
checking whether the version number is exactly matched before using any of the 
APIs?

And when are you going to release the wiimote version to the Android Market?

I apologize if I should have created another issue for these questions.
```

Original comment by `freeman....@gmail.com` on 7 Sep 2011 at 6:16

**Comment by GoogleCodeExporter** [comment]
```
I intend to keep the API fully backwards compatible. The version number is 
intended to allow applications to detect if a certain (new) feature is 
available. New features will likely be added either as a new broadcast message 
or an extra field in the package, and neither should introduce problems for 
apps that use the API.

I tried putting a version on Market with support for Wiimote, and a notice that 
this does not work for all phones. After a short period I got bad reviews 
simply stating "does not work". I am quite sure they just tried Wiimote and it 
did not work. Since the app works quite well with the other controllers, I do 
not want the Wiimote troubles to affect the overall app rating. By serving the 
Wiimote enabled app here, at least people have to read the Market text before 
they can download and try it. I have not found any solid way to detect if the 
OS has L2CAP, so I cannot do the detection reliably on the device.
```

Original comment by `kenneth@hexad.dk` on 8 Sep 2011 at 7:27

**Comment by GoogleCodeExporter** [comment]
```
I understand about bad reviews. So you probably will never put it back on the 
Market?
```

Original comment by `freeman....@gmail.com` on 8 Sep 2011 at 8:16

**Comment by GoogleCodeExporter** [comment]
```
As it looks now, no. 

I was considering having two versions, one with Wiimote and one without, but it 
gets complicated due to the broadcast messages, reviews, updates etc. If I 
either get L2CAP working or find a reliable way to detect if it is working, I 
will include the Wiimote driver in the Market version.


```

Original comment by `kenneth@hexad.dk` on 8 Sep 2011 at 8:20

**Comment by GoogleCodeExporter** [comment]
```
Oh, by the way, if you pass the driver string "wiimote" to BluezIME when 
connecting, it will use the Wiimote driver anyway. This means that if you have 
a custom controller bluetooth search and driver selection dialog, you can just 
add the Wiimote driver as an option, even if BluezIME does not report it.

The market version is built just like the version here, but I comment out the 
wiimote drivers in the top of BluezService.java, meaning that they cannot be 
selected and are not reported, but they work regardless (depending on the 
device though).
```

Original comment by `kenneth@hexad.dk` on 10 Sep 2011 at 11:00

**Comment by GoogleCodeExporter** [comment]
```
Cool! Thanks for the tip. Now we only need to fix issue #80 (it was actually 
submitted by my tester).
```

Original comment by `freeman....@gmail.com` on 10 Sep 2011 at 2:36

**Comment by GoogleCodeExporter** [comment]
```
I can see that you made it work, cool :)
```

Original comment by `kenneth@hexad.dk` on 3 Oct 2011 at 8:06
- Changed state: **Done**
